python/lsst/source/injection/inject_engine.py

```diff
@@ -236,16 +236,19 @@ def make_galsim_object(
     object_class = getattr(galsim, source_type)
     object_data = get_object_data(source_data, object_class)
     object = object_class(**object_data)
-    # Create a version of the object with an area-preserving shear applied.
-    shear_data = get_shear_data(source_data)
-    if shear_data:
-        try:
-            object = object.shear(**shear_data)
-        except TypeError as err:
-            if logger:
-                logger.warning("Cannot apply shear to GalSim object: %s", err)
-            pass
-    # Apply the instrumental flux and return.
+    # Don't shear delta functions. It will work if the parameters are finite
```

Comment:
This isn't the conclusion I thought we had reached on slack.

I'd rather not single out DeltaFunction if not necessary.  A better design would be to trap any GSObject with a nonsensical shear.  Something like adding the following within the try block:

```python
assert np.all(np.isifinite(object.jac)):
```

which would also trap `galsim.Gaussian(fwhm=1).shear(e1=np.nan)`, for example.

Reply:
I did not think we had actually reached a conclusion on Slack, because I'm not clear on what we think a 'nonsensical' shear is for a point source.

Reply:
Also, shearing point sources is unnecessary whether the shear is valid or not.

Reply:
I think a nonsensical shear is independent of whether or not the original profile is a point source.

My reading of the proposed change is that given the following table:

|type | e1 | e2| ... |
|-----|---|----|---|
| DeltaFunction | np.nan | 0.0 | ... |
| Gaussian | np.nan | 0.0| ... |
| VonKarman | np.nan | 0.0 | ... |

two sources get logged as bad (but the code continues) but the third goes through and gets injected into the image.  (I admit I didn't actually try to run this.  Happy to be corrected if I've misread here).  I'd strongly prefer that all three get logged as bad and not injected into the image.  I think that's the least surprising behavior.

Reply:
> I think a nonsensical shear is independent of whether or not the original profile is a point source.

I don't agree. Everything I'm planning to add is using the `q`, `beta` parameterization. As it is, the code will accept `0<q<=1` and any finite beta for a point source. But why? `DeltaFunctions` have an axis ratio of 0/0, so I'd say `nan` is the _only_ valid value.

At the very least it should accept both shear parameters being `nan` for a `DeltaFunction`, which I think can be checked easily. For other "invalid" values (let's say -99, which is sometimes used as a sentinel for values that should be positive) I'd prefer a single warning (per task, not per row), as not injecting should be a last resort if there's really no way to do it.

Having said all that, this can and probably should go on a different ticket since I don't strict need it to work as long as the other change to using `arrow_to_astropy` goes through.

Also, as a separate issue, `gs.Shear(q=np.nan, beta=gs.Angle(0, unit=gs.degrees))` does not raise (nor does `beta=gs.Angle(np.nan, unit=gs.degrees)`), and in fact the code continues and raises an uncaught `ValueError` when `drawImage` is called, which brings the whole task down - that's how I found this issue in the first place. So that needs to be fixed on another ticket either way.

Reply:
> `DeltaFunctions` have an axis ratio of 0/0, so I'd say nan is the only valid value.

Thanks, I think I see more of where you're coming from now.  That said, I think instead of as setting an object axis ratio / orientation, a more useful way to think of q/beta (or any shear parameterization) is that they apply a coordinate transformation (in such a way that circles become ellipses with the given axis ratio ...).  That's how GalSim treats all of this, and makes `DeltaFunction().shear(...)` make sense.  (And similarly, things like `Box().shear()` or `InterpolatedImage().shear()` where these profiles don't have an axis ratio to set).

How did we get here in the first place again?  IIRC, the nans appeared after vstacking a table of stars (with no q/beta) with a table of galaxies (with q/beta) results in masked values, and then roundtripping through parquet turned those masked values into nans?  Does parquet then introduce a risk of, for example, vstacking a q/beta catalog with an e1/e2 catalog resulting in passing 4 kwargs to `.shear()` (which is an error regardless of nan status)?  I suppose removing any nan-valued kwargs before applying .shear(**kwargs) or constructing a GSObject might be reasonable in this case in order to support parquet.

The fact you can construct galsim.Shears with np.nan does feel like a GalSim bug to me.

Reply:
I got here by making my own injection catalog out of DC2's truth_summary; it doesn't have anything to do with vstack behaviour. The bug is the way astropy reads parquet and converts masked values to nan (hence the necessary change below). I'm essentially advocating on behalf of hypothetical users who prefer to use nans over masked columns, though I will continue (attempting) to use masked columns myself.

The more general point is that users do not have to use any of the provided scripts to make or ingest injection catalogs, so while more validation there is fine, there does need to be more runtime validation. Any user can just butler import an ingestion catalog with whatever combination of columns they like. Ignoring both masked (as is done now) and nan-valued columns seems fine. You could then conceivably inject a catalog with say 4 shear columns so long as only 2 are provided per row. 

`galsim.Shear` only accepts up to 2 kwargs, although it has default values of `e1=0` and `e2=0` so unless explicitly checked, a catalog with no shear columns would work and inject circular sources (pre-convolution).

Reply:
It’s hard to say if this is a bug. A masked column in parquet has no values underlying the mask. It’s a feature of parquet.

Reply:
The bug isn't that it doesn't set the underlying value, it's that astropy turns masked parquet columns into unmasked columns with nans for the masked entries. Try:

```from lsst.daf.butler.formatters.parquet import arrow_to_astropy, astropy_to_arrow, pa, pq
from astropy.table import Table

filename = "/sdf/data/rubin/user/dtaranu/dc2/truth_summary_cell/truth_summary_v2_0_ci_imsim_2_2i_truth_summary.parq"

tab = arrow_to_astropy(pq.read_table(filename))
tab_ap = Table.read(filename)```

... at the USDF and you'll see what I mean.

Reply:
Chatting some more with @erykoff and the conclusion is that astropy's approach (which Eli wrote in the first place!) is consistent and not exactly wrong. If you write a masked column it saves it as such in parquet and loads it from file correctly. However, the butler formatter stores masked arrays as an unmasked array with `null` values as this is more storage-efficient in arrow. `astropy_to_arrow` and `arrow_to_astropy` also round-trip correctly but the problem is when using astropy to read a table written by `astropy_to_arrow` - it interprets unmasked arrays with nulls as an unmasked array with nans instead of a masked array.

I think it makes more sense to follow the stack/butler convention here and use `arrow_to_astropy`, and like I said earlier, I can drop this change and sort it out on another ticket.

Reply:
It seems to me like there are 3 issues being addressed on this ticket:

1. Using astropy to read a table written by `astropy_to_arrow` introduces `NaN` entries for every previously masked entries;
2. Shears are being applied to `DeltaFunction` types, which doesn't make sense;
3. Shears are not validated at runtime.

I think the solution for 1) on this ticket makes sense, and should be implemented. Please move the `# The character_as_bytes=False option is preferred, if possible.` in-line comment down to above the `try/except` block, and add a new in-line comment above the `arrow_to_astropy` call explaining why `.parquet` files are getting special treatment.

For 2), I agree with @jmeyers314 that singling out the `DeltaFunction` alone seems like not the way to go. As mentioned, there are other profile types for which this doesn't make sense either. Furthermore, if I understand correctly, addressing 1) will resolve 2) for your specific use case. 

For 3), I do like the idea of validating shears at runtime. Having an assert on the Jacobean right after the `try` statement will fail fast, and allow the user to re-assess their input catalog.